doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,411 @@
+- `KEY_FLAG_COW_NAME`: the name points to the name of another key
+- `KEY_FLAG_COW_META`: metakeys are copy-on-write
+
+A new copy flag will be added:
+
+- `KEY_CP_COW`: instructs `keyCopy` to copy whatever it should copy as copy-on-write.
+  This will NOT be part of `KEY_CP_ALL`.
+  We don't want developers outside of Elektra to accidentally use this.
+
+If `keyCopy()` is instructed to do a copy-on-write copy:
+
+- `dest->data.v` and `dest->data.c` point to the exact same location as in the source.
+  `dest->dataSize` is set to the same value as `source->dataSize`.
+  `KEY_FLAG_COW_VALUE` is set on `dest->flags`.
+  `KEY_FLAG_RO_VALUE` is set on `source->flags`.
+  `dest->originalData` is set to `source`.
+  `source->refs` is incremented.
+
+- `dest->key` points to `source->key`.
+  `dest->keySize` is set to the same value as `source->keySize`.
+  `dest->ukey` points to `source->ukey`.
+  `dest->keyUSize` is set to the same value as `source->keyUSize`.
+  `KEY_FLAG_COW_NAME` is set on `dest->flags`.
+  `KEY_FLAG_RO_NAME` is set on `source->flags`.
+  `dest->originalName` is set to `source`.
+  `source->refs` is incremented.
+
+- `dest->meta` points to a **new** keyset.
+  The keys in `dest->meta` are also copied with `KEY_CP_COW`, i.e. they are also copy-on-write keys.
+  `KEY_FLAG_COW_META` is set on `dest->flags`.
+  `KEY_FLAG_RO_META` is set on `source->flags`.
+
+The source key will remain as a read-only key.
+This constraint is needed, because the source key is the only key we can free the resources on.
+If the data or the name would change in the source key, all COW-copied keys would suddenly have another value.
+For the same reason, the source key will need to live longer than all COW-copied keys from it.
+
+A `keyCopy()` without `KEY_CP_COW` from an COW key will create a deep copy of the key.
+These keys are "normal" non-COW keys and can live on their own.
+
+Every `key*()` method that modifies data on a COW-copied key will need to allocate new memory for this data and remove the `KEY_FLAG_COW_DATA` flag.
+Every `key*()` method that modifies the name of a COW-copied key will need to allocate new memory for this name and remove the `KEY_FLAG_COW_NAME` flag.
+Every `key*()` method that modifies metadata needs to make sure that the same happens for metakeys.
+
+Keysets are not copy-on-write.
+A `ksDeepDup()` of a keyset with COW keys will create a keyset with deep-copied non-COW keys.
+Internally we may need a `ksCowDup()` function to create a keyset with copy-on-write keys from another keyset.
+Whether this function will be part of the public API is a point for discussion.
+
+### Full-blown copy-on-write implementation
+
+Make Elektra's `Key` and `KeySet` data structures copy-on-write.
+This requires some major refactoring of code within `libelektra-core`.
+Code that does only interact with the data structures via the public `libelektra-core` API should not notice any differences.
+The `mmapstorage` plugin will need a major refactoring.
```

Comment:
> The `mmapstorage` already needs to be fixed.

Why? What exactly is broken? AFAIK it works perfectly fine, [even as the default storage plugin](https://build.libelektra.org/blue/organizations/jenkins/libelektra/detail/master/858/pipeline/1062).

The `cache` plugin on the other hand, is broken and needs to be fixed because of the `new-backend` changes. Maybe that's what you're thinking of.

> **Note** `cache` and `mmapstorage` are separate plugins. `cache` depends on `mmapstorage`, but `mmapstorage` works without `cache`. Using the terms interchangeably is confusing and wrong.